iodata/formats/gamess.py

```diff
@@ -0,0 +1,145 @@
+# as published by the Free Software Foundation; either version 3
+# of the License, or (at your option) any later version.
+#
+# IODATA is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+# --
+"""GAMESS punch file format."""
+
+
+import numpy as np
+
+from ..docstrings import document_load_one
+from ..utils import angstrom, LineIterator
+
+
+__all__ = []
+
+
+PATTERNS = ['*.dat']
+
+
+def _read_data(lit: LineIterator) -> tuple:
+    """Extract ``title``, ``symmetry`` and ``symbols`` from the punch file."""
+    title = next(lit).strip()
+    symmetry = next(lit).split()[0]
+    # The dat file only contains symmetry-unique atoms, so we would be incapable of
+    # supporting non-C1 symmetry without significant additional coding.
+    if symmetry != "C1":
+        raise NotImplementedError(f"Only C1 symmetry is supported. Got {symmetry}")
+    symbols = []
+    line = True
+    while line != " $END      \n":
+        line = next(lit)
+        if line[0] != " ":
+            symbols.append(line.split()[0])
+    return title, symmetry, symbols
+
+
+def _read_coordinates(lit: LineIterator, result: dict) -> tuple:
+    """Extract ``numbers`` and ``coordinates`` from the punch file."""
+    for i in range(2):
+        next(lit)
+    natom = len(result["symbols"])
+    # if the data are already read before, just overwrite them
+    numbers = result.get("atnums")
+    if numbers is None:
+        numbers = np.zeros(natom, int)
+        result["atnums"] = np.zeros(natom, int)
+
+    coordinates = result.get("atcoords")
+    if coordinates is None:
+        coordinates = np.zeros((natom, 3), float)
+    for i in range(natom):
+        words = next(lit).split()
+        numbers[i] = int(float(words[1]))
+        coordinates[i] = np.array([float(elem) for elem in words[2:5]]) * angstrom
+    return numbers, coordinates
+
+
+def _read_energy(lit: LineIterator, result: dict) -> tuple:
+    """Extract ``energy`` and ``gradient`` from the punch file."""
+    energy = float(next(lit).split()[1])
+    natom = len(result["symbols"])
+    # if the data are already read before, just overwrite them
+    gradient = result.get("gradient")
+    if gradient is None:
+        gradient = np.zeros((natom, 3), float)
+    for i in range(natom):
+        words = next(lit).split()
+        gradient[i] = words[2:5]
+    return energy, gradient
+
+
+def _read_hessian(lit: LineIterator, result: dict) -> np.ndarray:
+    """Extract ``hessian`` from the punch file."""
+    # check that $HESS is not already parsed
+    if "athessian" in result:
+        lit.error("Cannot parse $HESS twice! Make sure approximate hessian is not being parsed!")
```

Comment:
I am not sure how to test this without adding a new test file (which is not desirable). Any suggestions @tovrstra?

Reply:
See other comment. This is as good as we can get it at the moment.